Added Data Access Objects for:

StanimirIliev opened the pull request.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcAccountRepository.kt:
```diff
@@ -0,0 +1,100 @@
+        return jdbcTemplate.execute("DELETE FROM $table WHERE Id=$accountId") == 1
+    }
+
+    override fun authenticate(accountId: Int, userId: Int): Boolean {
```
This looks strange. authenticate by just an accountId and userId ?

StanimirIliev replied:
My goal was to check if this account belongs to this user

StanimirIliev replied:
Maybe the name is inappropriate?

mgenov replied:
I've added this in another comment too: 

> When you go to specific account then you have to call: getUserAccount(userId, accountId)

e.g when you try to load account details for a given user you have to provide both and if it's not available a null account will be returned.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcSessionRepository.kt:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    private fun getRandomString(size: Int): String {
+        val random = Random()
```
UUID could be used instead: UUID.randomUUID()

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcSessionRepository.kt:
```diff
@@ -0,0 +1,89 @@
+        while (availableIds.contains(newId)) {// Get unique Id
+            newId = getRandomString(20)
+        }
+        session.expiresAt.nanos = 0//   mysql rounding issue
```
This means that expiresAt is not set from tests which doesn't look nice. To be able to easy test expiration of session you have to pass the time of which session was created and use some session time interval to set expiration time. 

For Example:

```kotlin
val repository = JdbcSessionRepository(template, "sessions", 60.minutes())

var session = repository.registerSession(Session(id="123",createdOn=LocalDateTime.of(2018, 1, 10, 8, 30, 0)))

// check session 5 minutes after 8:30 and it shouldn't be null
session = repository.getSession(session.id(), LocalDateTime.of(2018, 1, 10, 8, 35, 0)) 

// check session 1 day after the creation and it should be null
session = repository.getSession(session.id(), LocalDateTime.of(2018, 1, 11, 8, 30, 0)) 
```

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcSessionRepository.kt:
```diff
@@ -0,0 +1,89 @@
+        return activeSessionsCount
+    }
+
+    override fun getSessionId(userId: Int, expiresAt: Timestamp): String? {
```
This design doesn't look good. You won't be able to know when session expires as 

getSessionId requires userId to be passed but you know which is the user by the session Id. 

```kotlin
val session = repository.getNotAxpiredSessionAt(sessionId, instant: LocalDateTime) 
```

StanimirIliev replied:
Can one user have more than one sessions or when it logs in from other computer just the 'createdOn' DateTime will be updated of his session?

mgenov replied:
Yes, one user may have multiple sessions. 

1. Multiple browsers
2. Multiple computers

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcSessionRepository.kt:
```diff
@@ -0,0 +1,89 @@
+        return if(list.isEmpty()) null else list.first()
+    }
+
+    override fun deleteSession(sessionId: String): Boolean {
```
You need to delete periodically expired session after some time which means that this method should accept LocalDateTime 

```kotlin
override fun deleteSessionExpiredAfter(instant: LocalDateTime): Int /* number of deleted sessions */
```

StanimirIliev replied:
I need this method to delete sessions when users log outs. Can I make **deleteSessionsExpiredAfter** private in the implementation and to delete expired sessions when the method **getSession** is called?

mgenov replied:
It's not a good idea as you may have hundreds of sessions for deletions and this may slow down the request of the user. 

Instead of naming it delete, you can name it terminate or close 
terminateSession(sessionId: String)
closeSession(sessionId: String)

Delete sounds like a CRUD operation which doesn't sound good.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcTransactionRepository.kt:
```diff
@@ -0,0 +1,37 @@
+                })
+    }
+
+    override fun getTransactionId(userId: Int, accountId: Int, onDate: Timestamp): Int {
```
Doesn't look like a reasonable method. You can return the id of the transactions in the getTransactions method.

mgenov commented on backend/src/main/kotlin/com/clouway/app/core/Account.kt:
```diff
@@ -0,0 +1,25 @@
+package com.clouway.app.core
+
+class Account(val title: String, val userId: Int, val currency: Currency, val balance: Float, var id: Int = -1) {
```
data class generates equals and hashCode for you.

mgenov commented on backend/src/main/kotlin/com/clouway/app/core/AccountRepository.kt:
```diff
@@ -0,0 +1,74 @@
+     * @param userId the user id of account
+     * @return the id of this account or null if there is no account registered with these parameters
+     */
+    fun getAccountId(title: String, userId: Int): Int
```
Doesn't look like a reasonable method. 

When you go to the account lists you have to call getAllAccounts(userId)  which should returns all accounts and there ids. 

When you go to specific account then you have to call: getUserAccount(userId, accountId)

mgenov commented on backend/src/main/kotlin/com/clouway/app/core/AccountRepository.kt:
```diff
@@ -0,0 +1,74 @@
+     * @param userId the id of the user
+     * @return true if this account belongs to this user, false if does not
+     */
+    fun authenticate(accountId: Int, userId: Int): Boolean
```
This doesn't look like a reasonable function.

mgenov commented on backend/src/main/kotlin/com/clouway/app/core/SessionRepository.kt:
```diff
@@ -0,0 +1,41 @@
+package com.clouway.app.core
+
+import java.sql.Timestamp
```
Domain object shouldn't depend on SQL package as it's part of the adapter. 

User LocalDateTime instead as it provides an easy setup in kotlin.

mgenov replied:
If JDBC requires Timestamp, then you have to adapt it in the Jdbc implementation.

mgenov commented on backend/src/main/kotlin/com/clouway/app/core/Transaction.kt:
```diff
@@ -0,0 +1,28 @@
+
+import java.sql.Timestamp
+
+class Transaction(val userId: Int, val accountId: Int, val onDate: Timestamp, val operation: Operation,
```
data class provides equals and hashCode for you.

StanimirIliev commented:
Can you take a look again please

mgenov commented on backend/build.gradle:
```diff
@@ -22,15 +38,21 @@ dependencies {
     compile group: 'mysql', name: 'mysql-connector-java', version: '6.0.6'
-    compile group: 'org.eclipse.jetty', name: 'jetty-server', version: '9.4.7.v20170914'
-    compile group: 'org.eclipse.jetty', name: 'jetty-servlet', version: '9.4.7.v20170914'
+//    compile group: 'org.eclipse.jetty', name: 'jetty-server', version: '9.4.7.v20170914'
```
Commented stuff could be removed.

mgenov commented on backend/src/main/kotlin/com/clouway/app/App.kt:
```diff
@@ -1,5 +1,5 @@
 package com.clouway.app
 
 fun main(args: Array<String>) {
```
Looks un-used

StanimirIliev replied:
Should I remove it?

mgenov replied:
Yes, if it's not used.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcAccountRepository.kt:
```diff
@@ -0,0 +1,88 @@
+                })
+    }
+
+    override fun removeAccount(accountId: Int): Boolean {
```
UserId need to be provided as I'll be able to remove accounts of other users.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcAccountRepository.kt:
```diff
@@ -0,0 +1,88 @@
+        return if (list.isEmpty()) null else list.first()
+    }
+
+    private fun getUserId(accountId: Int): Int? {
```
Doesn't look like a reasonable method.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcAccountRepository.kt:
```diff
@@ -0,0 +1,88 @@
+        return if (affectedRows != 1 || list.isEmpty()) -1 else list.first()
+    }
+
+    override fun updateBalance(accountId: Int, amount: Float): Boolean {
```
User id should be passed here and it should come from the session of the user.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcAccountRepository.kt:
```diff
@@ -0,0 +1,88 @@
+        if (amount + balance < 0) {
+            return false
+        }
+        if (amount == 0f) {
```
This looks strange. updateBalance function is not doing anything when amount is zero ?

StanimirIliev replied:
Yes, when the user decide to deposit 0 money, there is no need to update anything

StanimirIliev commented:
I made changes

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcSessionRepository.kt:
```diff
@@ -0,0 +1,61 @@
+                }).count()
+    }
+
+    override fun getSessionId(userId: Int, createdOn: LocalDateTime): String? {
```
I'm not sure that you need such a method as sessions are used for retrieving of userId.

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcSessionRepository.kt:
```diff
@@ -0,0 +1,61 @@
+        private val table: String,
+        private val userTable: String
+) : SessionRepository {
+    override fun registerSession(session: Session): String? {
```
The expiration time of session is not available in the insert query. The expiration time is required as you are not able to check whether the session was expired or not.

```kotlin
val session = repository.getSessionAvailableAt(sessionId, instant: LocalDateTime) 
```

StanimirIliev commented:
I made changes:   
* added *expiresAt* property in the Session DTO
* added parameter *instant* for the methods *getSessionsCount* and *getSessionAvailableAt*
* the *UserRepository's* method *registerUser* now returns the id of the new registered user
* removed *UserRepository's* method *getUserId*

StanimirIliev commented:
I made changes:   
* added *expiresAt* property in the Session DTO
* added parameter *instant* for the methods *getSessionsCount* and *getSessionAvailableAt*
* the *UserRepository's* method *registerUser* now returns the id of the new registered user
* removed *UserRepository's* method *getUserId*

mgenov commented on backend/src/test/kotlin/com/clouway/app/adapter/jdbc/JdbcAccountRepositoryTest.kt:
```diff
@@ -0,0 +1,165 @@
+        transactionRepository = JdbcTransactionRepository(mySqlJdbcTemplate, transactionTable)
+        accountRepository = JdbcAccountRepository(mySqlJdbcTemplate, transactionRepository, accountsTable)
+        userRepository = JdbcUserRepository(mySqlJdbcTemplate, userTable)
+        val statement = mySqlDataSource.connection.createStatement()
```
This code looks duplicated between tests. You can move it into a datastore rule which reads all tables and truncates them. The simple method probably would be to erase specific list of tables as you are able to control the truncate/delete order. 

```kotlin
@Rule @JvmField
val datastoreRule = DatastoreRule()
```

mgenov commented on backend/src/main/kotlin/com/clouway/app/adapter/jdbc/JdbcUserRepository.kt:
```diff
@@ -1,32 +1,60 @@
+        return if (list.isEmpty()) false else list.first()
+    }
+
+    override fun getUsername(id: Int): String? {
```
getUsername looks used only in tests which means that it could be removed.

mgenov commented on backend/schema/Accounts.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE Accounts (
```
Seems complicated every DDL to be in separate files. You can move everything into bank_ddl.sql

StanimirIliev commented:
Can you take another look please
